packages/cmd/agent.go

```diff
@@ -157,44 +461,123 @@ func (d *DynamicSecretLeaseManager) Append(lease DynamicSecretLease) {
+func (d *DynamicSecretLeaseManager) GetLease(accessToken, projectSlug, environment, secretPath, slug string, templateId int) *DynamicSecretLease {
 	d.mutex.Lock()
 	defer d.mutex.Unlock()
 
+	// first try to get from in-memory storage
+
 	for _, lease := range d.leases {
-		if lease.SecretPath == secretPath && lease.Environment == environment && lease.ProjectSlug == projectSlug && lease.Slug == slug && slices.Contains(lease.TemplateIDs, templateId) {
+		if lease.SecretPath == secretPath && lease.Environment == environment && lease.ProjectSlug == projectSlug && lease.Slug == slug && lease.TemplateID == templateId {
 			return &lease
 		}
 	}
 
+	// if no lease is found in in-memory storage, try to get from cache
+
+	log.Info().Msgf("[cache]: no lease found, fetching from cache")
+	leaseFromCache := d.ReadLeaseFromCache(projectSlug, environment, secretPath, slug, templateId)
+	log.Debug().Msgf("[cache]: lease from cache: %+v", leaseFromCache)
+
+	if leaseFromCache != nil {
+
+		// try to get the lease from the API
+
+		// ? question(daniel): should we trust the cache more, and avoid calling the API to see if the lease still exists?
+		// ? we do this to ensure that the lease wasn't deleted while the agent was not running
+
+		dynamicSecretLease, err := util.GetDynamicSecretLease(accessToken, leaseFromCache.ProjectSlug, leaseFromCache.Environment, leaseFromCache.SecretPath, leaseFromCache.LeaseID)
+		if err != nil {
+
+			log.Warn().Msgf("[cache]: error: %+v", err)
+
+			// lease not found in API, delete it from cache and return nil
+			if errors.Is(err, api.ErrNotFound) {
+				log.Warn().Msgf("dynamic secret lease does not exist, deleting from cache: [lease-id=%s]", leaseFromCache.LeaseID)
+				if err := d.DeleteLeaseFromCache(leaseFromCache.ProjectSlug, leaseFromCache.Environment, leaseFromCache.SecretPath, leaseFromCache.Slug, leaseFromCache.TemplateID); err != nil {
+					log.Warn().Msgf("[cache]: unable to delete lease from cache: %v", err)
+				}
+
+				return nil
+			}
+
+			// lease is found in cache but not in the the API, and the API returned a non 404-error. We should attempt to revoke it
+			// at this point we know that we should be able to reach the API because we've done authentication successfully
+			log.Warn().Msgf("unable to get dynamic secret lease from API. Revoking lease from cache: [lease-id=%s]", leaseFromCache.LeaseID)
+			if err := d.DeleteLeaseFromCache(leaseFromCache.ProjectSlug, leaseFromCache.Environment, leaseFromCache.SecretPath, leaseFromCache.Slug, leaseFromCache.TemplateID); err != nil {
+				log.Warn().Msgf("[cache]: unable to delete lease from cache: %v", err)
+			}
+
+			if err := revokeDynamicSecretLease(accessToken, leaseFromCache.ProjectSlug, leaseFromCache.Environment, leaseFromCache.SecretPath, leaseFromCache.LeaseID); err != nil {
+				log.Warn().Msgf("unable to revoke dynamic secret lease %s: %v", leaseFromCache.LeaseID, err)
+				return nil
+			}
```

Comment:
We are deleting it anyway, so the if should be to revoke or not:

```suggestion
				if err := d.DeleteLeaseFromCache(leaseFromCache.ProjectSlug, leaseFromCache.Environment, leaseFromCache.SecretPath, leaseFromCache.Slug, leaseFromCache.TemplateID); err != nil {
					log.Warn().Msgf("[cache]: unable to delete lease from cache: %v", err)
				}

            // only attempt to revoke if the lease exists (not a 404 error)
			// if it's 404, the lease doesn't exist so there's nothing to revoke
			if !errors.Is(err, api.ErrNotFound) {
				log.Warn().Msgf("unable to get dynamic secret lease from API. Attempting to revoke lease: [lease-id=%s]", leaseFromCache.LeaseID)
				if err := revokeDynamicSecretLease(accessToken, leaseFromCache.ProjectSlug, leaseFromCache.Environment, leaseFromCache.SecretPath, leaseFromCache.LeaseID); err != nil {
					log.Warn().Msgf("unable to revoke dynamic secret lease %s: %v", leaseFromCache.LeaseID, err)
				}
			} else {
				log.Warn().Msgf("dynamic secret lease does not exist in API: [lease-id=%s]", leaseFromCache.LeaseID)
			}
```

Reply:
Not sure I'm following, what we want is to is:

If 404:
Delete from cache, don't even try to revoke. Just return right away

If not 404:
Delete from cache, and try to revoke


I feel like both are trying to achieve the same, but I feel like what we have currently reads a bit easier?